tests/cpp/kvstore/gpu_topology_test.cc

```diff
@@ -0,0 +1,672 @@
+  for (unsigned i = 0; i < scan.size(); ++i)
+    ASSERT_EQ(scan[i], correct_scan[i]);
+}
+
+// UpdateWeightTest
+TEST(GpuTopology, TestUpdateWeight) {
+  std::vector<float> W = {0.f, 1.f,
+                          1.f, 0.f};
+  std::vector<size_t> topo = {1, 1, 0};
+  int num_gpus = 2;
+  float alpha  = 0.7;
+  std::vector<float> correct_W = {0.f, 0.7f,
+                                  0.7f, 0.f};
+  mxnet::kvstore::UpdateWeight(&W, topo, num_gpus, alpha);
+  ASSERT_EQ(W.size(), correct_W.size());
+  for (unsigned i = 0; i < W.size(); ++i) {
+    ASSERT_EQ(W[i], correct_W[i]);
+  }
+}
+
+// ComputeTreesFromRoot
+TEST(GpuTopology, TestComputeTreesFromRoot1) {
+  std::vector<float> W = {0, 2, 2, 3, 3, 1, 1, 1,
+                          2, 0, 3, 2, 1, 3, 1, 1,
+                          2, 3, 0, 3, 1, 1, 2, 1,
+                          3, 2, 3, 0, 1, 1, 1, 2,
+                          3, 1, 1, 1, 0, 2, 2, 3,
+                          1, 3, 1, 1, 2, 0, 3, 2,
+                          1, 1, 2, 1, 2, 3, 0, 3,
+                          1, 1, 1, 2, 3, 2, 3, 0};
+  int num_gpus = 8;
+  int root     = 0;
+  float alpha  = 0.7;
+  bool backtrack = true;
+  unsigned correct_topo_size = 15;
+  unsigned correct_scan_size = 5;
+  std::vector<size_t> topo;
+  std::vector<size_t> scan;
+
+  mxnet::kvstore::ComputeTreesFromRoot(&W, num_gpus, root, alpha, backtrack,
+                                       &topo, &scan);
+
+  ASSERT_EQ(topo.size(), correct_topo_size);
+  ASSERT_EQ(scan.size(), correct_scan_size);
+}
+
+// IsConnected
+// Test on graph that is "disconnected" by NVLink
+TEST(GpuTopology, TestIsConnected1) {
+  std::vector<float> W = {0, 0, 2, 0,
+                          0, 0, 0, 2,
+                          2, 0, 0, 0,
+                          0, 2, 0, 0};
+  int num_gpus = 4;
+
+  bool connected = mxnet::kvstore::IsConnected(W, num_gpus);
+
+  bool correct_connected = false;
+  ASSERT_EQ(connected, correct_connected);
+}
+
+// IsConnected
+// Test on graph that is "disconnected" by NVLink
+TEST(GpuTopology, TestIsConnected2) {
+  std::vector<float> W = {1, 1, 2, 1,
+                          1, 1, 1, 2,
+                          2, 1, 1, 1,
+                          1, 2, 1, 1};
+  int num_gpus = 4;
+
+  bool connected = mxnet::kvstore::IsConnected(W, num_gpus);
+
+  bool correct_connected = false;
+  ASSERT_EQ(connected, correct_connected);
+}
+
+// IsConnected
+// Test on graph that is "disconnected" by NVLink
+TEST(GpuTopology, TestIsConnected3) {
+  std::vector<float> W = {1, 1, 2, 2,
+                          1, 1, 1, 2,
+                          2, 1, 1, 1,
+                          2, 2, 1, 1};
+  int num_gpus = 4;
+
+  bool connected = mxnet::kvstore::IsConnected(W, num_gpus);
+
+  bool correct_connected = true;
+  ASSERT_EQ(connected, correct_connected);
+}
+
+// ComputeTreesTest with backtracking
+// TODO(carlyang): comment out test for now
+/*TEST(GpuTopology, TestComputeTrees1) {
```

Comment:
What's wrong with these tests? Do they not work?

Reply:
They used to segfault only on CI, but now they should be fine. I fixed an off-by-1 bug.